### PR TITLE
airbyte-ci: upgrade to dagger 0.9.5

### LINKED ISSUE #33429: airbyte-ci: upgrade to dagger 0.9.5
We're currently pinning version 0.6.4 of dagger on airbyte-ci and in our infra.

Upgrading to the latest dagger version would be beneficial for:
* stability: bug fixes and improvements have been shipped since 0.6.4, especially on the distributed cache feature. **Using the latest dagger version could make us use the Dagger hosted magicache that we currently host ourself.**
* support: when facing bugs / outages it's easier to get support from Dagger team if we're running their latest version
* new features: Dagger latest version has new feature like external networking, service orchestration etc. that we could leverage to improve airbyte-ci dind use cases. It also starts supporting "re-usable dagger components" a.k.a Dagger Modules / Zenith.

## Infra actions to take:
1. Declare a new `RunnerDeployment` variant [here](https://github.com/airbytehq/airbyte-infra/blob/main/manifests/runner-deployment.yml#L903) : `ci-runner-connector-dev-large-dagger-0-9-4`
2. Declare a new `Provisioner`variant [here](https://github.com/airbytehq/airbyte-infra/blob/main/manifests/karpenter.yml#L825) like `ci-runner-connector-dev-large-dagger-0-6-4`
3. Declare a new Daemonset variant [here](https://github.com/airbytehq/airbyte-infra/blob/main/manifests/dagger-daemonset-remote-cache-conn.yml#L154): like `connector-dev-dagger-0-9-4-engine`: set the image to `registry.dagger.io/engine:v0.9.4` Set the `nodeAffinity` rules to target `ci-runner-connector-dev-large-dagger-0-9-4`. Please note that the `_EXPERIMENTAL_*` env var should be discarded and replaced with different env var (e.g. `DAGGER_CLOUD_TOKEN`). We should checkout Dagger docs / sync with their team to make sure we assign the correct env vars to the engine.
4. `kubectl` apply the changes manifest

## Testing out the new version by changing the client code **on a branch**:
* Update dagger to 0.9.4 in `pipelines/pyproject.toml`
* Changes `runs_on` of the workflow you'd like to test to `ci-runner-connector-dev-large-dagger-0-9-4`
* Fix any breaking changes

## Note about Dagger Cloud tokens:
**Dagger cloud token should now be considered a secret as they are now used for cache syncs**. We should remove the public occurrences of the hardcoded token: `p.eyJ1IjogIjFiZjEwMmRjLWYyZmQtNDVhNi1iNzM1LTgxNzI1NGFkZDU2ZiIsICJpZCI6ICJlNjk3YzZiYy0yMDhiLTRlMTktODBjZC0yNjIyNGI3ZDBjMDEifQ.hT6eMOYt3KZgNoVGNYI3_v4CC-s19z8uQsBkGrBhU3k`. The token to use is available on the Dagger Cloud console.

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Closes #33429

## How
* Bump Dagger to 0.9.5 in: `airbyte-ci`, `base-images`, `CAT`
* Update code where broken due to backward-incompatible APIs
* Update GHA workflows to use the 0.9.5 engines on the infra

## Manual testing
(manual testing should be done using the [dev airbyte ci binary](https://storage.googleapis.com/dev-airbyte-cloud-connector-metadata-service/airbyte-ci/releases/ubuntu/8af28fc/airbyte-ci) - this is why I used [skip ci] on all commits)
- [x]  [`connectors --name=source-postgres --name=source-faker test` ](https://github.com/airbytehq/airbyte/actions/runs/7398615148/job/20128139667)
- [x] [`connectors --name=source-pokeapi publish --pre-release`](https://github.com/airbytehq/airbyte/actions/runs/7398665671)
- [x] [`format check all`](https://github.com/airbytehq/airbyte/actions/runs/7398492384/job/20127744709?pr=33582)
- [x] `format fix all` (local)
- [x] `connectors --name=source-faker bump_version` (local)
- [x] [nightly builds](https://github.com/airbytehq/airbyte/actions/runs/7398648409/job/20128248526)
- [x] `airbyte-ci test`
## 🚨 User Impact 🚨
None, if all manual tests succeed

## Dependencies:
1. **This should not be merged until the `magicache` set up is complete on the infra side.**
**Update**: Distributed caching is currently enabled on our runners: waiting for Dagger team to confirm it's working correctly.

2 **This should not be merged until CI can run on current branche's binary**
Depends on https://github.com/airbytehq/airbyte/pull/33519

## Related infra PRs
* https://github.com/airbytehq/airbyte-infra/pull/87
* https://github.com/airbytehq/airbyte-infra/pull/88

### CODE REVIEW

**Comment by alafanechere** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #33582** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/33582?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/airbytehq/airbyte/33582?utm_source=stack-comment).

File: airbyte-ci/connectors/pipelines/pyproject.toml [hunk 1]
```diff
@@ -10,7 +10,7 @@ authors = ["Airbyte <contact@airbyte.io>"]
 
 [tool.poetry.dependencies]
 python = "~3.10"
-dagger-io = "^0.6.4"
+dagger-io = "==0.9.5"
```
Comment (postamar) [inline airbyte-ci/connectors/pipelines/pyproject.toml:13]: Shouldn't you bump the major version counter of airbyte-ci as well?

File: airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py [hunk 2]
```diff
@@ -140,13 +140,13 @@ def with_global_dockerd_service(
...
 
-    dockerd_container = dockerd_container.with_new_file("/etc/docker/daemon.json", daemon_config_json)
+    dockerd_container = dockerd_container.with_new_file("/etc/docker/daemon.json", contents=daemon_config_json)
     if docker_hub_username_secret and docker_hub_password_secret:
         # Docker login happens late because there's a cache buster in the docker login command.
         dockerd_container = docker_login(dockerd_container, docker_hub_username_secret, docker_hub_password_secret)
     return dockerd_container.with_exec(
         ["dockerd", "--log-level=error", f"--host=tcp://0.0.0.0:{DOCKER_HOST_PORT}", "--tls=false"], insecure_root_capabilities=True
-    )
+    ).as_service()
```
Comment (postamar) [inline airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py:149]: neat

**Review by postamar — APPROVED** [review]
Does this mean we can follow this up with a PR which removes the dirty hack I put in that installs some secrets as regular env vars with `--is-local` is set so that the gradle integration test output wouldn't be all messed up?

File: .github/actions/run-dagger-pipeline/action.yml [hunk 3]
```diff
@@ -142,14 +145,14 @@ runs:
         export _EXPERIMENTAL_DAGGER_RUNNER_HOST="unix:///var/run/buildkit/buildkitd.sock"
         airbyte-ci --disable-dagger-run --is-ci --gha-workflow-run-id=${{ github.run_id }} ${{ inputs.subcommand }} ${{ inputs.options }}
       env:
-        _EXPERIMENTAL_DAGGER_CLOUD_TOKEN: "p.eyJ1IjogIjFiZjEwMmRjLWYyZmQtNDVhNi1iNzM1LTgxNzI1NGFkZDU2ZiIsICJpZCI6ICJlNjk3YzZiYy0yMDhiLTRlMTktODBjZC0yNjIyNGI3ZDBjMDEifQ.hT6eMOYt3KZgNoVGNYI3_v4CC-s19z8uQsBkGrBhU3k"
```
Comment (bnchrch) [inline .github/actions/run-dagger-pipeline/action.yml:145]: ❤️ Nice to see this go to an input!

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/pipeline.py [hunk 4]
```diff
@@ -94,11 +94,6 @@ async def run_connectors_pipelines(
     default_connectors_semaphore = anyio.Semaphore(concurrency)
     dagger_logs_output = sys.stderr if not dagger_logs_path else create_and_open_file(dagger_logs_path)
     async with dagger.Connection(Config(log_output=dagger_logs_output, execute_timeout=execute_timeout)) as dagger_client:
-
-        # HACK: This is to get a long running dockerd service to be shared across all the connectors pipelines
-        # Using the "normal" service binding leads to restart of dockerd during pipeline run that can cause corrupted docker state
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/pipeline.py:99]: 😍

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/pipeline.py [hunk 5]
```diff
@@ -109,27 +104,21 @@ async def run_connectors_pipelines(
...
-                    context.dockerd_service = dockerd_service
-                    tg_connectors.start_soon(
-                        connector_pipeline,
-                        context,
-                        CONNECTOR_LANGUAGE_TO_FORCED_CONCURRENCY_MAPPING.get(context.connector.language, default_connectors_semaphore),
-                        *args,
-                    )
-            # When the connectors pipelines are done, we can stop the dockerd service
-            tg_main.cancel_scope.cancel()
+        await dockerd_service.start()
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/pipeline.py:107]: 💅 nit: Should/could we turn this to an async with

```py
async with dockerd_service.start():
```

File: airbyte-ci/connectors/pipelines/pipelines/dagger/actions/connector/hooks.py [hunk 6]
```diff
@@ -3,16 +3,23 @@
...
 import importlib.util
+from importlib import metadata
 from importlib.abc import Loader
 
 from dagger import Container
-from dagger.engine._version import CLI_VERSION as dagger_engine_version
 from pipelines.airbyte_ci.connectors.context import ConnectorContext
 
 
+def get_sdk_version() -> str:
```
Comment (bnchrch) [inline airbyte-ci/connectors/pipelines/pipelines/dagger/actions/connector/hooks.py:13]: Not immediately clear what SDK its refering to when you call it.
```suggestion
def get_dagger_sdk_version() -> str:
```

**Review by bnchrch — APPROVED** [review]
A few very small things but LGTM!

File: airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/pipeline.py (hunk 5 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/connectors/pipeline.py:107]: It'd be nice but I don't think we can as it's not returning a context manager

File: airbyte-ci/connectors/pipelines/pipelines/dagger/actions/connector/hooks.py (hunk 6 above)
Comment (alafanechere) [reply airbyte-ci/connectors/pipelines/pipelines/dagger/actions/connector/hooks.py]: Done

**Comment by alafanechere** [comment]
> Does this mean we can follow this up with a PR which removes the dirty hack I put in that installs some secrets as regular env vars with --is-local is set so that the gradle integration test output wouldn't be all messed up?

@postamar I think it's worth a try, if my not mistaken the issues you opened on Dagger repo are closed now .

**Pull request merged**